theatre/studio/src/utils/invariant.ts

```diff
@@ -0,0 +1,72 @@
+import {stringify} from './stringify'
+
+type accepts = string | number | object
+export function invariant(
+  cond: any,
+  /** Pass a message or function producing a message */
+  message: (() => accepts) | accepts,
+  /** Pass a value for what we found instead */
+  found?: any,
+): asserts cond {
+  if (!cond) {
+    throw new InvariantError(
+      'Invariant: ' +
+        stringify(typeof message === 'function' ? message() : message) +
+        (arguments.length > 2 ? `\nInstead found: ${stringify(found)}` : ''),
+      found,
+    )
+  }
+}
+/** Plain old throw with a never return to help the type system */
+export function invariantThrow(
+  /** Pass a message or function producing a message */
+  message: (() => accepts) | accepts,
+  /** Pass a value for what we found instead */
+  found?: any,
+): never {
+  throw new InvariantError(
+    'Invariant: ' +
+      stringify(typeof message === 'function' ? message() : message) +
+      (arguments.length > 1
+        ? `\nInstead found: ${
+            // check if probably already contains stringified json
+            /": "/.test(found) ? found : stringify(found)
+          }`
+        : ''),
+    found,
```

Comment:
Why is this code so similar but slightly different to the body of `invariant`? Can it be refactored?  Can you pull out some of these expressions into variables with expressive names?

Reply:
We might be able to add another creator function to InvariantError, but generally, the rationale is to ensure that the stack trace is kept minimal.